components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/OIDCScopeHandler.java

```diff
@@ -49,15 +51,30 @@ public boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx) throws Id
             // if id_token is allowed for requested grant type.
             return true;
         } else {
+
             // Remove openid scope from the token message context.
             String[] scopes = (String[]) ArrayUtils.removeElement(tokReqMsgCtx.getScope(), OAuthConstants.Scope.OPENID);
+
+            //Get all the registered oidc scopes of a tenant
+            List<String> oidcScopes = OAuth2Util.getOIDCScopes(tokReqMsgCtx.getOauth2AccessTokenReqDTO()
+                    .getTenantDomain());
+
+            //Check whether the is there any oidc scope in the request
+            for (String scope : scopes) {
+                if (oidcScopes.contains(scope)) {
+                    scopes = (String[]) ArrayUtils.removeElement(scopes, scope);
+                }
+            }
```

Comment:
Can we improve this logic to filter the OIDC scopes at once without repeatedly removing the oidc scopes one by one?